### ISSUE TITLE
BUG: NDFrame.__setattr__ calls object.__setattribute__ on the same name, which can cause unexpected behavior (rewrote #56773)

### ISSUE THREAD

**Issue by dhodczak** [pandas-dev/pandas#56793]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```
python
from pandas import DataFrame
from functools import cached_property
class Frame(DataFrame):
    @cached_property
    def spam(self):
        print('<time wasting task>')

Frame().spam = 5
```

```
<time wasting task>
```

### Issue Description

In the example, the user is trying to assign a value to a property, but `NDFrame.__setattr__` first calls `object.__getattribute__` on the same name. It wastes runtime by unnecessarily calling properties or raising `AttributeError`. Here is the original code:
```
    @final
    def __setattr__(self, name: str, value) -> None:
        """
        After regular attribute access, try setting the name
        This allows simpler access to columns for interactive use.
        """
        # first try regular attribute access via __getattribute__, so that
        # e.g. ``obj.x`` and ``obj.x = 4`` will always reference/modify
        # the same attribute.

        try:
            object.__getattribute__(self, name)
            return object.__setattr__(self, name, value)
        except AttributeError:
            pass

        # if this fails, go on to more involved attribute setting
        # (note that this matches __getattr__, above).
        if name in self._internal_names_set:
            object.__setattr__(self, name, value)
        elif name in self._metadata:
            object.__setattr__(self, name, value)
        else:
            try:
                existing = getattr(self, name)
                if isinstance(existing, Index):
                    object.__setattr__(self, name, value)
                elif name in self._info_axis:
                    self[name] = value
                else:
                    object.__setattr__(self, name, value)
            except (AttributeError, TypeError):
                if isinstance(self, ABCDataFrame) and (is_list_like(value)):
                    warnings.warn(
                        "Pandas doesn't allow columns to be "
                        "created via a new attribute name - see "
                        "https://pandas.pydata.org/pandas-docs/"
                        "stable/indexing.html#attribute-access",
                        stacklevel=find_stack_level(),
                    )
                object.__setattr__(self, name, value)


```

### Expected Behavior

The attribute should not be accessed, avoiding otherwise wasted time or unexpected behavior. 

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 0f437949513225922d851e9581723d82120684a6
python           : 3.11.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 6.1.69-1-MANJARO
Version          : #1 SMP PREEMPT_DYNAMIC Thu Dec 21 12:29:38 UTC 2023
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 2.0.3
numpy            : 1.24.4
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 65.5.1
pip              : 23.2.1
Cython           : 0.29.36
pytest           : 7.4.0
hypothesis       : None
sphinx           : 7.1.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.14.0
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.2
numba            : 0.57.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.11.1
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : 0.9.0
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

**Comment by rhshadrach** [comment]
@dhodcz2 - is this the same as #56773?

**Comment by dhodcz2** [comment]
Yes, that was my first issue posted and kind of bad--I rewrote it as it's more of a "bug" than a "feature request," thanks

**Comment by jbrockmendel** [comment]
Is there any evidence in the git blame about why it is done this way?  Do any tests fail if we remove the check?

**Comment by jbrockmendel** [comment]
This broke 50ish tests when I tried removing it this morning.

**Comment by rhshadrach** [comment]
We could check for a property / cached_property by looking at attributes on `type(self)`, but I feel like we might end up playing wack-a-mole. Users can do all sorts of odd things by defining their own `__getattribute__`, but perhaps that's sufficiently esoteric we shouldn't worry about it, whereas property / cached_property are more common.

**Comment by jbrockmendel** [comment]
Closing as no action.
